FLIP 56: Non-Fungible Token Standard Version 2 (#56)
onflow/flips: application/20221219-nft-v2.md

---
status: Implemented 
flip: 56

> authors: Joshua Hannan (joshua.hannan@dapperlabs.com)
> sponsor: Joshua Hannan (joshua.hannan@dapperlabs.com) 

Reviewer: ```suggestion
authors: Joshua Hannan (joshua.hannan@flowfoundation.org)
sponsor: Joshua Hannan (joshua.hannan@flowfoundation.org) 
```

The current token standard uses contract interfaces and nested type requirements.
They are designed in a way that requires each concrete contract
to provide exactly one `NFT` and one `Collection` type.
This means that any project that needs multiple tokens must deploy multiple contracts.
In the case of very simple tokens, this is a lot of complexity for very little value.

Related to this problem, functionality and metadata associated with some tokens,
such as paths and empty vault creation methods
is only accessible directly through the contract itself, when it should also
be accessible directly through an instance of a token resource and/or interface.

Many contracts also do not implement MetadataViews properly because
there is no requirement for it in the standard.

In the case of events, currently there is no way for the standard to ensure that
implementations are emitting standardized and correct events, which this upgrade will address.

## User Benefit

With these upgrades, users will now be able to:
* Define multiple tokens in a single contract.
* Query all the data about a token directly through the contract as well as the `Vault` resource and core interfaces.

> * Have standard events emitted correctly for important operations (`Withdrawn`, `Deposited`, `Updated`, and `NonFungibleToken.NFT.ResourceDestroyed`)

Reviewer 2: nit, add a ending period to be consisten.

is on the Flow Forum.
A [pull request with the suggested code changes](https://github.com/onflow/flow-nft/pull/126) 
is in the flow fungible token github repository.

The main code changes and their implications are described here.
The linked proposals provide more context, though the proposal in the forum post
is quite out of date.

### Add support for multiple token types in a single contract

Now that nested type requirements have been removed, contracts can define
multiple token types. The type information can be accessed by using the built-in
`contract.publicTypes(): [Type]` method. A return value of this is now required
in the `NonFungibleToken.createEmptyCollection(nftType: Type)` method because
the caller needs to specify which type they are looking for.

`ViewResolver` has also been updated to require type information
when querying views from a contract:

```cadence
access(all) contract interface ViewResolver {

    /// Function that returns all the Metadata Views implemented by the resolving contract.
    /// Some contracts may have multiple resource types that support metadata views

>     /// so there there is an optional parameter for specify which resource type the caller

Reviewer 2: there there, double word

    /// is looking for views for.
    /// Some contract-level views may be type-agnostic. In that case, the contract
    /// should return the same views regardless of what type is passed in.
    ///
    /// @param resourceType: An optional resource type to return views for
    /// @return An array of Types defining the implemented views. This value will be used by
    ///         developers to know which parameter to pass to the resolveView() method.
    ///
    access(all) view fun getContractViews(resourceType: Type?): [Type]

    /// Function that resolves a metadata view for this token.
    /// Some contracts may have multiple resource types that support metadata views

>     /// so there there is an optional parameter for specify which resource type the caller
>     /// is looking for views for.

Reviewer 3: ```suggestion
    /// so there there is an optional parameter to specify which resource type the caller
    /// is requesting views for.
```

```

### Add `getAcceptedTypes()` method to `Receiver` interface

It is useful to be able to query a non-fungible token receiver to see what types of tokens
it can accept.

```cadence
access(all) resource interface Receiver {
    /// getSupportedNFTTypes returns a list of NFT types that this receiver accepts
    /// @return A dictionary of types mapped to booleans indicating if this
    ///         reciever supports it
    access(all) view fun getSupportedNFTTypes(): {Type: Bool}

    /// Returns whether or not the given type is accepted by the collection
    /// A collection that can accept any type should just return true by default
    /// @param type: An NFT type
    /// @return A boolean indicating if this receiver can recieve the desired NFT type
    access(all) view fun isSupportedNFTType(type: Type): Bool
}
```

### Change the return type of `borrowNFT()` to an optional

In the current standard, `borrowNFT()` is meant to panic if the NFT to borrow
is not found in the collection. This is clearly the wrong choice, because 
it is a best practice to have getter functions return `nil` if something goes wrong

> instead of panicing and reverting the transaction.

Reviewer 2: panicing, should be "panicking" just spelling.

```cadence

> pub fun borrowNFT(id: UInt64): &AnyResource{NFT}?

Reviewer 3: ```suggestion
access(all) view fun borrowNFT(_ id: UInt64): &{NFT}?
```

```

### Remove the requirement for the `ownedNFTs` field in `Collection`

The requirement to include an `ownedNFTs: {UInt64: @NFT}` field
in Collection implementations is restrictive because developers may want
to store tokens in a unique way that is different what what could be specified.
This proposal removes `ownedNFTs` and only requires a `getIDs()` method
for the collection to indicate which IDs it contains.

> ### Move `createEmptyCollection()` to inside the Collection definition in addition to the contract

Reviewer 3: Might be worth mentioning this is also added to the `NFT` interface as well for similar reasoning.

It is useful to be able to create a new empty collection
directly from a `Collection` object instead of having to import
the contract and call the method from the contract.

```cadence

> pub resource interface Collection {

Reviewer 3: ```suggestion
access(all) resource interface Collection {
```

    /// createEmptyCollection creates an empty Collection
    /// and returns it to the caller so that they can own NFTs
    access(all) fun createEmptyCollection(): @{Collection} {
        post {
            result.getType() == self.getType(): "The created collection does not have the same type as this collection"
            result.getLength() == 0: "The created collection must be empty!"
        }
    }
}
```

### Add Metadata Views methods to the standard

Metadata Views for non-fungible tokens should be easily accessible
from interfaces defined by the standard. 

> This propsal enforces that implementations implement the `ViewResolver` contract interface
> and `NFT` implementations implement the `ViewResolver.Resolver` interface.

Reviewer 3: ```suggestion
This proposal enforces that implementations implement the `ViewResolver` contract interface, `NFT` implementations implement the `ViewResolver.Resolver` interface, and `Collection` implementations implement the `ViewResolver.ResolverCollection` interface.
```

* The code can be tested on its own once a compatible version of cadence is released. 

### Best Practices

* Some of the changes illustrate Cadence best practices, such as encapsulating functionality
within resources, avoiding public fields, and giving developers flexibility to write composable code.

### Tutorials and Examples

* Coming soon once the proposal reaches a more final state.

### Compatibility

* FCL, emulator, and other such tools should not be affected besides potentially
having to update standard transactions if they aren't compatible.

### User Impact

* The upgrade will go out at the same time as stable cadence if approved

## Related Issues

### Scoped Providers

A critical piece of tooling for non-fungible tokens would be a struct that contains
a provider capability but restricts the capability to only be able to withdraw
specified IDs or number of tokens from the underlying collection.
Currently, providers have no limit, but all tokens should be able
to create scoped providers.

This feature is out of the scope of this proposal, but should definitely be a standard
that lives alongside the main fungible token standard.

> We hope to shepard a proposal for these soon.

Reviewer 3: ```suggestion
We hope to shepherd a proposal for these soon.
```
